code/controllers/subsystems/serialization.dm

```diff
@@ -0,0 +1,316 @@
+		return
+	
+	var/datum/object = new object_type()
+	var/address = data[NAMEOF(src, persistent_address)]
+//	object.load_deserialized_data(data, options)
+	object.iterate_deserialized_variables(data, options)
+	deserialized_objects[address] = weakref(object)
+	most_recent_deserialized_data = data.Copy()
+//	process_reference_lookup_queue()
+	return object
+
+// File I/O
+
+// Writes a json string to disk, storing it permanently.
+// This proc will happily overwrite whatever is in the path, so be aware of that.
+/datum/controller/subsystem/serialization/proc/write_json(json_string, file_path)
+	var/json_file = file(file_path)
+	if(fexists(json_file))
+		fdel(json_file) // This stops the file from getting appended, and breaking the json parser.
+	//to_file(json_file, json_string)
+	text2file(json_string, json_file)
+
+// Reads a json file from disk, retriving the json string contained inside.
+/datum/controller/subsystem/serialization/proc/read_json(file_path)
+	if(!fexists(file_path))
+		CRASH("Tried to open file '[file_path]', but it does not exist or cannot be accessed.")
+	
+	var/json_file = file(file_path)
+	return file2text(json_file)
+
+// Adds some whitespace to make the json formatting easier for humans to read.
+// `json_decode()` ignores it so it's purely for the benefit of humans.
+// However it will add a small amount of overhead when something is being saved.
+// The file size will also be slightly bigger due to the added whitespace.
+/datum/controller/subsystem/serialization/proc/pretty_json(json_string)
+	// Remove existing whitespace, to avoid double pretty-ifying the string if it contains a json string
+	// that was already processed by this proc.
+	json_string = replacetext(json_string, "\n", "")
+	json_string = replacetext(json_string, "\\n", "")
+//	json_string = replacetext(json_string, " ", "")
+
+	var/depth = 0 // How much to indent.
+	var/inside_string = FALSE // If it sees an unescaped `"`, don't do anything.
+	// This is being done as a list in order to reduce repeatitive concatination, which helps preserve BYOND's super secret string tree.
+	var/list/new_json_string_list = list() 
+	for(var/i = 1 to length(json_string))
+		var/char = copytext(json_string, i, i+1)
+		var/string_to_add = char
+		switch(char)
+			if("\"")
+				// If encountering an open or closing apostrophe, check if it's an escaped one. 
+				if(!(copytext(json_string, i-1, i) == "\\"))
+					inside_string = !inside_string
+			if("{", "\[")
+				if(!inside_string)
+					depth++
+					string_to_add = "[char]\n[indent(depth)]"
+			if("}", "\]")
+				if(!inside_string)
+					depth--
+					string_to_add = "\n[indent(depth)][char]"
+			if(",")
+				if(!inside_string)
+					string_to_add = "[char]\n[indent(depth)]"
+		new_json_string_list += string_to_add
+	return new_json_string_list.Join()
+
+// Used for above proc.
+/datum/controller/subsystem/serialization/proc/indent(amount)
```

Comment:
`pad_right()` or `pad_left()` on Neb might be worth porting as they're useful procs in general.